// @langchain/langgraph-supervisor/dist/handoff.cjs
const require_rolldown_runtime = require('./_virtual/rolldown_runtime.cjs');
const __langchain_langgraph = require_rolldown_runtime.__toESM(require("@langchain/langgraph"));
const uuid = require_rolldown_runtime.__toESM(require("uuid"));
const zod = require_rolldown_runtime.__toESM(require("zod"));
const __langchain_core_messages = require_rolldown_runtime.__toESM(require("@langchain/core/messages"));
const __langchain_core_tools = require_rolldown_runtime.__toESM(require("@langchain/core/tools"));

//#region src/handoff.ts
const WHITESPACE_RE = /\s+/;
function _normalizeAgentName(agentName) {
	/**
	* Normalize an agent name to be used inside the tool name.
	*/
	return agentName.trim().replace(WHITESPACE_RE, "_").toLowerCase();
}
const createHandoffTool = ({ agentName, agentDescription }) => {
	/**
	* Create a tool that can handoff control to the requested agent.
	*
	* @param agentName - The name of the agent to handoff control to, i.e.
	*   the name of the agent node in the multi-agent graph.
	*   Agent names should be simple, clear and unique, preferably in snake_case,
	*   although you are only limited to the names accepted by LangGraph
	*   nodes as well as the tool names accepted by LLM providers
	*   (the tool name will look like this: `transfer_to_<agent_name>`).
	*/
	const toolName = `transfer_to_${_normalizeAgentName(agentName)}`;
	const handoffTool = (0, __langchain_core_tools.tool)(async (_, config) => {
		/**
		* Ask another agent for help.
		*/
		const toolMessage = new __langchain_core_messages.ToolMessage({
			content: `Successfully transferred to ${agentName}`,
			name: toolName,
			tool_call_id: config.toolCall.id
		});
		const state = (0, __langchain_langgraph.getCurrentTaskInput)();
		return new __langchain_langgraph.Command({
			goto: agentName,
			graph: __langchain_langgraph.Command.PARENT,
			update: { messages: state.messages.concat(toolMessage) }
		});
	}, {
		name: toolName,
		schema: zod.z.object({}),
		description: agentDescription ?? "Ask another agent for help."
	});
	return handoffTool;
};
function createHandoffBackMessages(agentName, supervisorName) {
	/**
	* Create a pair of (AIMessage, ToolMessage) to add to the message history when returning control to the supervisor.
	*/
	const toolCallId = (0, uuid.v4)();
	const toolName = `transfer_back_to_${_normalizeAgentName(supervisorName)}`;
	const toolCalls = [{
		name: toolName,
		args: {},
		id: toolCallId
	}];
	return [new __langchain_core_messages.AIMessage({
		content: `Transferring back to ${supervisorName}`,
		tool_calls: toolCalls,
		name: agentName
	}), new __langchain_core_messages.ToolMessage({
		content: `Successfully transferred back to ${supervisorName}`,
		name: toolName,
		tool_call_id: toolCallId
	})];
}

//#endregion
exports.createHandoffBackMessages = createHandoffBackMessages;
exports.createHandoffTool = createHandoffTool;
