// @langchain/langgraph-supervisor/dist/supervisor.js
import { createHandoffBackMessages, createHandoffTool } from "./handoff.js";
import { START, StateGraph } from "@langchain/langgraph";
import { createReactAgent, createReactAgentAnnotation, withAgentName } from "@langchain/langgraph/prebuilt";
import { v5 } from "uuid";

//#region src/supervisor.ts
const PROVIDERS_WITH_PARALLEL_TOOL_CALLS_PARAM = new Set(["ChatOpenAI"]);
function isChatModelWithBindTools(llm) {
	return "_modelType" in llm && typeof llm._modelType === "function" && llm._modelType() === "base_chat_model" && "bindTools" in llm && typeof llm.bindTools === "function";
}
function isChatModelWithParallelToolCallsParam(llm) {
	return llm.bindTools.length >= 2;
}
function isRemoteGraph(agent) {
	if (agent == null || typeof agent !== "object") return false;
	if (!("lc_id" in agent)) return false;
	if (!Array.isArray(agent.lc_id)) return false;
	return agent.lc_id.join(".") === "langgraph.pregel.RemoteGraph";
}
const makeCallAgent = (agent, outputMode, addHandoffBackMessages, supervisorName) => {
	if (!["full_history", "last_message"].includes(outputMode)) throw new Error(`Invalid agent output mode: ${outputMode}. Needs to be one of ["full_history", "last_message"]`);
	return async (state, config) => {
		let conf = config;
		if (isRemoteGraph(agent)) {
			const threadId = config?.configurable?.thread_id;
			const agentThreadId = threadId && agent.name ? v5(agent.name, threadId) : null;
			conf = {
				...config ?? {},
				configurable: {
					...config?.configurable ?? {},
					thread_id: agentThreadId
				}
			};
		}
		const output = await agent.invoke(state, conf);
		let { messages } = output;
		if (outputMode === "last_message") messages = messages.slice(-1);
		if (addHandoffBackMessages) messages.push(...createHandoffBackMessages(agent.name, supervisorName));
		return {
			...output,
			messages
		};
	};
};
/**
* Create a multi-agent supervisor.
*
* @param params Parameters for the supervisor.
* @returns The supervisor graph.
*/
const createSupervisor = ({ agents, llm, tools, prompt, responseFormat, stateSchema, contextSchema, outputMode = "last_message", addHandoffBackMessages = true, supervisorName = "supervisor", includeAgentName, preModelHook, postModelHook }) => {
	const agentNames = /* @__PURE__ */ new Set();
	for (const agent of agents) {
		if (!agent.name || agent.name === "LangGraph") throw new Error("Please specify a name when you create your agent, either via `createReactAgent({ ..., name: agentName })` or via `graph.compile({ name: agentName })`.");
		if (agentNames.has(agent.name)) throw new Error(`Agent with name '${agent.name}' already exists. Agent names must be unique.`);
		agentNames.add(agent.name);
	}
	const handoffTools = agents.map((agent) => {
		const agentName = agent.name;
		const agentDescription = "description" in agent && typeof agent.description === "string" ? agent.description : void 0;
		return createHandoffTool({
			agentName,
			agentDescription
		});
	});
	const allTools = [...tools ?? [], ...handoffTools];
	let supervisorLLM = llm;
	if (isChatModelWithBindTools(llm)) {
		if (isChatModelWithParallelToolCallsParam(llm) && PROVIDERS_WITH_PARALLEL_TOOL_CALLS_PARAM.has(llm.getName())) supervisorLLM = llm.bindTools(allTools, { parallel_tool_calls: false });
		else supervisorLLM = llm.bindTools(allTools);
		supervisorLLM.kwargs ??= {};
		const kwargs = supervisorLLM.kwargs;
		if (!("tools" in kwargs)) {
			if ("config" in supervisorLLM && typeof supervisorLLM.config === "object" && supervisorLLM.config != null && "tools" in supervisorLLM.config) kwargs.tools = supervisorLLM.config.tools;
		}
	}
	if (includeAgentName) supervisorLLM = withAgentName(supervisorLLM, includeAgentName);
	const schema = stateSchema ?? createReactAgentAnnotation();
	const supervisorAgent = createReactAgent({
		name: supervisorName,
		llm: supervisorLLM,
		tools: allTools,
		prompt,
		responseFormat,
		stateSchema: schema,
		preModelHook,
		postModelHook
	});
	let builder = new StateGraph(schema, contextSchema).addNode(supervisorAgent.name, supervisorAgent, { ends: [...agentNames] }).addEdge(START, supervisorAgent.name);
	for (const agent of agents) {
		builder = builder.addNode(agent.name, makeCallAgent(agent, outputMode, addHandoffBackMessages, supervisorName), { subgraphs: isRemoteGraph(agent) ? void 0 : [agent] });
		builder = builder.addEdge(agent.name, supervisorAgent.name);
	}
	return builder;
};

//#endregion
export { createSupervisor, withAgentName };
